task_manager/task_manager.hh

```diff
@@ -0,0 +1,323 @@
+        utils::updateable_value<uint32_t> task_ttl = utils::updateable_value<uint32_t>(0);
+    };
+    using task_ptr = seastar::shared_ptr<task_manager::task>;
+    using task_map = std::unordered_map<task_id, task_ptr>;
+    using foreign_task_ptr = seastar::foreign_ptr<task_ptr>;
+    using foreign_task_vector = std::vector<foreign_task_ptr>;
+    using module_ptr = seastar::shared_ptr<module>;
+    using modules = std::unordered_map<std::string, module_ptr>;
+private:
+    task_map _all_tasks;
+    modules _modules;
+    config _cfg;
+    abort_source& _as;
+    serialized_action _update_task_ttl_action;
+    utils::observer<uint32_t> _task_ttl_observer;
+    uint32_t _task_ttl;
+public:
+    class task_not_found : public std::exception {
+        sstring _cause;
+    public:
+        explicit task_not_found(task_id tid)
+            : _cause(format("task with id {} not found", tid.to_sstring()))
+        { }
+
+        virtual const char* what() const noexcept override { return _cause.c_str(); }
+    };
+    enum class task_state {
+        created,
+        running,
+        done,
+        failed
+    };
+    class task {
+    public:
+        struct progress {
+            double completed = 0.0;         // Number of units completed so far.
+            double total = 0.0;             // Total number of units to complete the task.
+        };
+        struct status {
+            task_id id;
+            std::string type;
+            task_state state = task_state::created;
+            db_clock::time_point start_time;
+            db_clock::time_point end_time;
+            std::string error;
+            int sequence_number = 0;        // A running sequence number of the task.
+            unsigned shard = 0;
+            std::string keyspace;
+            std::string table;
+            std::string entity;             // Additional entity specific for the given type of task.
+            std::string progress_units;     // A description of the units progress.
+        };
+    protected:
+        status _status;
+        progress _progress;                 // Reliable only for tasks with no descendants.
+        foreign_task_ptr _parent;
+        foreign_task_vector _children;
+        seastar::shared_promise<> _done;
+        module_ptr _module;
+    public:
+        task(module_ptr module, task_id id, unsigned sequence_number, std::string keyspace, std::string table, std::string type, std::string entity, foreign_task_ptr&& parent) 
+            : _module(module) {
+            _status.id = id;
+            _status.state = task_state::created;
+            _status.shard = this_shard_id();
+            _status.sequence_number = sequence_number;
+            _status.keyspace = std::move(keyspace);
+            _status.table = std::move(table);
+            _status.type = std::move(type);
+            _status.entity = std::move(entity);
+            if (parent.get() != nullptr) {
+                _parent = std::move(parent);
```

Comment:
hmm, why not move it unconditionally then?

Reply:
If you initialize _status in the initialization list (using designated initializers for the different members), you can then initialize _parent as well there.